## Thông tin sinh viên

- Họ và tên: Trần Hoài Thiện Nhân
- MSSV: 23127238
- Lớp: 23KHDL

## Import các thư viện cần thiết

In [1]:
import datetime as dt # Để làm việc với dữ liệu datetime
import math

## Đưa ra câu hỏi cần trả lời

Tình trạng hiểu về Python các bạn trong lớp?

## Thu thập dữ liệu

Cho làm quiz về Python trên moodle.

Kết quả có thể được download về dưới dạng file .csv.

## Khám phá dữ liệu

In [2]:
file = open('Data/PythonQuiz_CLC_2025.csv', 'r',encoding='utf-8-sig')

cols = {}
first_line_vals = file.readline().rstrip().split(',')
for first_line_val in first_line_vals:
    cols[first_line_val] = []
for line in file:
    line_vals = line.rstrip().split(',')
    if (line_vals[0] == 'In progress' or line_vals[0] == ''):
        continue
    for i in range(len(line_vals)):
        cols[first_line_vals[i]].append(line_vals[i])

file.close()

cols

{'State': ['Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished'],
 'Started on': ['24/10/2025 17:39',
  '24/10/2025 18:09',
  '24/10/2025 18:34',
  '24/10/2025 19:07',
  '24/10/2025 19:55',
  '24/10/2025 20:11',
  '24/10/2025 21:30',
  '24/10/2025 22:09',
  '24/10/2025 22:13',
  '24/10/2025 22:17',
  '24/10/2025 22:26',
  '25/

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [3]:
cols.keys()

dict_keys(['State', 'Started on', 'Completed', 'Time taken', 'Grade/10.00', 'Q. 1 /0.53', 'Q. 2 /0.53', 'Q. 3 /0.53', 'Q. 4 /0.53', 'Q. 5 /0.53', 'Q. 6 /0.53', 'Q. 7 /0.53', 'Q. 8 /0.53', 'Q. 9 /0.53', 'Q. 10 /0.53', 'Q. 11 /0.53', 'Q. 12 /0.53', 'Q. 13 /0.53', 'Q. 14 /0.53', 'Q. 15 /0.53', 'Q. 16 /0.53', 'Q. 17 /0.53', 'Q. 18 /0.53', 'Q. 19 /0.53'])

In [4]:
# Số dòng
len(cols['State'])

52

In [5]:
# Số cột
len(cols)

24

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng ứng với kết quả làm quiz của một sinh viên.

Xem qua bảng dữ liệu của file csv thì có vẻ không thấy có dòng nào "lạc loài" (thật ra dữ liệu ban đầu là có dòng cuối là dòng tính điểm trung bình của các cột, nhưng mình đã bỏ đi rồi).

### Dữ liệu có các dòng bị lặp không?

Xem qua bảng dữ liệu của file csv thì có vẻ không thấy có vấn đề một sinh viên xuất hiện nhiều hơn một lần.

### Mỗi cột có ý nghĩa gì?

- Cột "State": cho biết trạng thái làm bài, chẳng hạn "Finished" là đã nộp bài (đã ấn nút "submit all and finish")
- 3 cột "Started on", "Completed", "Time taken": lần lượt cho biết thời điểm bắt đầu làm, thời điểm làm xong, và thời gian từ khi bắt đầu làm cho tới khi làm xong
- Cột "Grade/10.00": cho biết điểm tổng (tối đa là 10, vì có 16 câu và mỗi câu 0.63 điểm)
- 19 cột còn lại "Q. 1 /0.63", "Q. 2 /0.63", ...: cho biết điểm của mỗi câu quiz (mỗi câu làm đúng hoàn toàn thì sẽ được 0.53 điểm)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [6]:
for col_name, col_vals in cols.items():
    print(f'{col_name} {type(col_vals[0])}')

State <class 'str'>
Started on <class 'str'>
Completed <class 'str'>
Time taken <class 'str'>
Grade/10.00 <class 'str'>
Q. 1 /0.53 <class 'str'>
Q. 2 /0.53 <class 'str'>
Q. 3 /0.53 <class 'str'>
Q. 4 /0.53 <class 'str'>
Q. 5 /0.53 <class 'str'>
Q. 6 /0.53 <class 'str'>
Q. 7 /0.53 <class 'str'>
Q. 8 /0.53 <class 'str'>
Q. 9 /0.53 <class 'str'>
Q. 10 /0.53 <class 'str'>
Q. 11 /0.53 <class 'str'>
Q. 12 /0.53 <class 'str'>
Q. 13 /0.53 <class 'str'>
Q. 14 /0.53 <class 'str'>
Q. 15 /0.53 <class 'str'>
Q. 16 /0.53 <class 'str'>
Q. 17 /0.53 <class 'str'>
Q. 18 /0.53 <class 'str'>
Q. 19 /0.53 <class 'str'>


Đúng như dự đoán, tất cả các cột đều đang có kiểu dữ liệu là str. Để có thể tiếp tục khám phá tiếp thì ta cần tiền xử lý để các cột có kiểu dữ liệu phù hợp.

- Cột "State": string là đúng rồi, không cần chuyển
- Cột "Started on" và "Completed on": chuyển sang kiểu dữ liệu datetime
- Cột "Time taken": chuyển sang một con số thể hiện cho số phút
- Các cột "Grade/10.00", "Q. 1 /0.63", ...: chuyển sang kiểu dữ liệu số

## Tiền xử lý

In [7]:
new_cols = {}
for col_name, col_value in cols.items():
    if '/' in col_name:
        new_col_val = []
        for val in col_value:
            try:
                f_val = float(val)
            except:
                f_val = 0
            new_col_val.append(f_val)
        new_cols[col_name] = new_col_val
    elif col_name in ['Started on','Completed']:
        new_col_val = []
        for val in col_value:
            d_value = dt.datetime.strptime(val,'%d/%m/%Y %H:%M')
            new_col_val.append(d_value)
        new_cols[col_name] = new_col_val
    elif col_name == 'Time taken':
        new_col_val = []
        for val in col_value:
            hour = 0;minute=0;secs=0;day=0;
            ls_value = val.split(' ')
            for i in range(len(ls_value)):
                if 'day' in ls_value[i]:
                    day = int(ls_value[i-1])
                elif 'hour' in ls_value[i]:
                    hour = int(ls_value[i-1])
                elif 'min' in ls_value[i]:
                    minute = int(ls_value[i-1])
                elif 'sec' in ls_value[i]:
                    secs = int(ls_value[i-1])
            new_value = (day * 24 + hour) * 60 + minute + secs/60
            new_col_val.append(new_value)
        new_cols[col_name] = new_col_val
    else:
        new_cols[col_name] = col_value
new_cols

{'State': ['Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished'],
 'Started on': [datetime.datetime(2025, 10, 24, 17, 39),
  datetime.datetime(2025, 10, 24, 18, 9),
  datetime.datetime(2025, 10, 24, 18, 34),
  datetime.datetime(2025, 10, 24, 19, 7),
  datetime.datetime(2025, 10, 24, 19, 55),
  datetime.datetime(2025, 10, 24, 2

In [8]:
cols = new_cols

In [9]:
for col_name, col_vals in cols.items():
    print(f'{col_name} {type(col_vals[0])}')

State <class 'str'>
Started on <class 'datetime.datetime'>
Completed <class 'datetime.datetime'>
Time taken <class 'float'>
Grade/10.00 <class 'float'>
Q. 1 /0.53 <class 'float'>
Q. 2 /0.53 <class 'float'>
Q. 3 /0.53 <class 'float'>
Q. 4 /0.53 <class 'float'>
Q. 5 /0.53 <class 'float'>
Q. 6 /0.53 <class 'float'>
Q. 7 /0.53 <class 'float'>
Q. 8 /0.53 <class 'float'>
Q. 9 /0.53 <class 'float'>
Q. 10 /0.53 <class 'float'>
Q. 11 /0.53 <class 'float'>
Q. 12 /0.53 <class 'float'>
Q. 13 /0.53 <class 'float'>
Q. 14 /0.53 <class 'float'>
Q. 15 /0.53 <class 'float'>
Q. 16 /0.53 <class 'float'>
Q. 17 /0.53 <class 'float'>
Q. 18 /0.53 <class 'float'>
Q. 19 /0.53 <class 'float'>


## Khám phá dữ liệu

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

- Số-lượng/tỉ-lệ các giá trị thiếu?
- Min? max? $\to$ Có gì bất thường không?

Ta thấy: cột "Started on" và "Completed" có kiểu dữ liệu datetime nhưng cũng có thể cho gia nhập hội numeric và tính min & max luôn.

Số-lượng/tỉ-lệ các giá trị thiếu? Không có giá trị thiếu.

Min? Max?

In [10]:
for col_name, col_vals in cols.items():
    if type(col_vals[0]) != str:
        print(f'{col_name}\t{min(col_vals)}\t{max(col_vals)}')

Started on	2025-10-24 17:39:00	2025-10-30 23:49:00
Completed	2025-10-24 18:31:00	2025-10-30 23:55:00
Time taken	6.6	4560.0
Grade/10.00	3.03	10.0
Q. 1 /0.53	0.0	0.53
Q. 2 /0.53	0.0	0.53
Q. 3 /0.53	0.0	0.53
Q. 4 /0.53	0.0	0.53
Q. 5 /0.53	0.0	0.53
Q. 6 /0.53	0.0	0.53
Q. 7 /0.53	0.0	0.53
Q. 8 /0.53	0.13	0.53
Q. 9 /0.53	0.26	0.53
Q. 10 /0.53	0.13	0.53
Q. 11 /0.53	0.26	0.53
Q. 12 /0.53	0.0	0.53
Q. 13 /0.53	0.0	0.53
Q. 14 /0.53	0.0	0.53
Q. 15 /0.53	0.0	0.53
Q. 16 /0.53	0.0	0.53
Q. 17 /0.53	0.0	0.53
Q. 18 /0.53	0.0	0.53
Q. 19 /0.53	0.0	0.53


Có vẻ không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

- Số-lượng/tỉ-lệ các giá trị thiếu?
- Số lượng các giá trị khác nhau? Show một vài giá trị \
$\to$ Có gì bất thường không?

Hội này chỉ có một thành viên là cột "State"!

Số-lượng/tỉ-lệ các giá trị thiếu? Không có giá trị thiếu.

Số lượng các giá trị khác nhau? Show một vài giá trị

In [11]:
for col_name, col_vals in cols.items():
    if type(col_vals[0]) == str:
        col_vals_set = set(col_vals)
        print(f'{col_name}\t{len(col_vals_set)}\t{col_vals_set}')

State	1	{'Finished'}


Có vẻ không có gì bất thường.

## Đưa ra câu hỏi cần trả lời

Sau khi đã hiểu hơn về dữ liệu, ta sẽ quay lại bước đưa ra câu hỏi cần trả lời: tinh chỉnh câu hỏi, hoặc có thể đưa ra thêm các câu hỏi mới.

Câu hỏi “tình trạng hiểu về Python của các bạn trong lớp?” có thể được cụ thể hóa ra thành 2 câu hỏi:

1. Cột điểm tổng “Grade/10.00” có phân bố các giá trị như thế nào?
2. Câu quiz nào có nhiều bạn làm sai nhất, câu nào có nhiều bạn làm sai nhì, ...?

## Tiền xử lý + phân tích dữ liệu để trả lời cho câu hỏi 1

*Câu hỏi 1: Cột điểm tổng “Grade/10.00” có phân bố các giá trị như thế nào?*

Ở đây không cần tiền xử lý mà sẽ đi luôn vào phân tích dữ liệu.

Đầu tiên, ta sẽ xem thử median của cột "Grade/10.00".

In [12]:
grades = cols['Grade/10.00']
sorted_grades = sorted(grades)
p = 50
i = int(p/100 * (len(sorted_grades)-1))
sorted_grades[i]

8.95

Để cảm thụ rõ hơn về phân bố các giá trị của cột "Grade/10.00", ta sẽ chia miền giá trị của cột này thành các khoảng (bin) và xem thử mỗi khoảng có bao nhiêu sinh viên.

Giả sử ta sẽ chia đoạn [0, 10] thành 5 bin như sau: [0, 2), [2, 4), [4, 6), [6, 8),[8,10]. Ta có thể lưu kết quả vào một dictionary, trong đó: key là một tuple cho biết 2 đầu của một bin (thay tuple bằng list thì có được không?), value là số lượng sinh viên rơi vào bin này.

***Sinh viên hoàn thành tiếp phần này như bài tập về nhà***

Thay `tuple` bằng `list` thì có được không? Trả lời trực tiếp vào cell này

$\to$ Không được. `tuple` có thể dùng làm key của `dictionary` vì `tuple` là immutable (không thay đổi được), còn `list` là mutable (có thể thay đổi) nên không thể dùng làm key của dictionary.

In [13]:
bins = {
    (0, 2): 0,
    (2, 4): 0,
    (4, 6): 0,
    (6, 8): 0,
    (8, 10): 0
}

for grade in grades:
    if 0 <= grade < 2:
        bins[(0, 2)] += 1
    elif 2 <= grade < 4:
        bins[(2, 4)] += 1
    elif 4 <= grade < 6:
        bins[(4, 6)] += 1
    elif 6 <= grade < 8:
        bins[(6, 8)] += 1
    elif 8 <= grade <= 10:
        bins[(8, 10)] += 1

print("Phân bố điểm theo khoảng:")
for (start, end), count in bins.items():
    percentage = (count / len(grades)) * 100
    if end == 10:
        print(f"[{start}-{end}] điểm: {count} sinh viên ({percentage:.3f}%)")
    else:
        print(f"[{start}-{end}) điểm: {count} sinh viên ({percentage:.3f}%)")


Phân bố điểm theo khoảng:
[0-2) điểm: 0 sinh viên (0.000%)
[2-4) điểm: 1 sinh viên (1.923%)
[4-6) điểm: 3 sinh viên (5.769%)
[6-8) điểm: 11 sinh viên (21.154%)
[8-10] điểm: 37 sinh viên (71.154%)


## Tiền xử lý + phân tích dữ liệu để trả lời cho câu hỏi 2

*Câu hỏi 2: Câu quiz nào có nhiều bạn làm sai nhất, câu nào có nhiều bạn làm sai nhì, ...?*

Ta có thể lưu kết quả vào một dictionary, trong đó: key là tên của cột ứng với câu quiz, value là điểm trung bình của cột. Dictionary này được sort tăng dần theo value.

***Sinh viên hoàn thành tiếp phần này như bài tập về nhà***

In [14]:
# Câu nào mà có điểm trung bình kiếm được ít nhất thì là câu khó nhất.

question_avg = {}

for col_name, col_vals in cols.items():
    if col_name.startswith('Q.'):
        total = 0
        for val in col_vals:
            total += val
        avg = total / len(col_vals)
        question_avg[col_name] = avg

sorted_questions = sorted(question_avg.items(), key=lambda x: x[1])

print("Thứ tự câu hỏi từ khó nhất đến dễ nhất theo điểm trung bình kiếm được:")
for i, (question, avg_score) in enumerate(sorted_questions, 1):
    print(f"{i}. {question}: {avg_score:.3f} điểm")

Thứ tự câu hỏi từ khó nhất đến dễ nhất theo điểm trung bình kiếm được:
1. Q. 6 /0.53: 0.357 điểm
2. Q. 5 /0.53: 0.377 điểm
3. Q. 18 /0.53: 0.387 điểm
4. Q. 4 /0.53: 0.398 điểm
5. Q. 13 /0.53: 0.418 điểm
6. Q. 14 /0.53: 0.418 điểm
7. Q. 1 /0.53: 0.426 điểm
8. Q. 3 /0.53: 0.428 điểm
9. Q. 17 /0.53: 0.438 điểm
10. Q. 19 /0.53: 0.448 điểm
11. Q. 7 /0.53: 0.459 điểm
12. Q. 15 /0.53: 0.469 điểm
13. Q. 8 /0.53: 0.475 điểm
14. Q. 10 /0.53: 0.491 điểm
15. Q. 11 /0.53: 0.494 điểm
16. Q. 9 /0.53: 0.509 điểm
17. Q. 2 /0.53: 0.510 điểm
18. Q. 12 /0.53: 0.515 điểm
19. Q. 16 /0.53: 0.520 điểm


## Kết luận 1:

**Câu khó nhất (nhiều bạn làm sai nhất - câu có điểm trung bình kiếm được ít nhất): Q. 6 /0.53**

**Câu dễ nhất (ít bạn làm sai nhất - câu có điểm trung bình kiếm được nhiều nhất): Q. 16 /0.53**

---

In [15]:
# Câu nào mà nhiều người bị điểm 0 nhất là sai nhiều nhất.

question_stats = {}

for col_name, col_vals in cols.items():
    if col_name.startswith('Q.'):
        count_zero = 0 
        count_full = 0
        
        for val in col_vals:
            if val == 0:
                count_zero += 1
            elif val == 0.53:
                count_full += 1
        
        total = len(col_vals)
        zero_percent = (count_zero / total) * 100
        full_percent = (count_full / total) * 100
        
        question_stats[col_name] = {
            'zero_count': count_zero,
            'full_count': count_full,
            'total': total,
            'zero_percent': zero_percent,
            'full_percent': full_percent
        }

sorted_questions = sorted(question_stats.items(), key=lambda x: x[1]['zero_count'], reverse=True)

print("Danh sách câu hỏi từ khó nhất đến dễ nhất")
print("(Dựa trên số người làm sai - 0 điểm)")

for i, (question, stats) in enumerate(sorted_questions, 1):
    print(f"{i}. {question}:")
    print(f"Số người sai (0 điểm): {stats['zero_count']}/{stats['total']} ({stats['zero_percent']:.3f}%)")
    print(f"Số người đúng hoàn toàn (5.3 điểm): {stats['full_count']}/{stats['total']} ({stats['full_percent']:.3f}%)")
    print()


Danh sách câu hỏi từ khó nhất đến dễ nhất
(Dựa trên số người làm sai - 0 điểm)
1. Q. 6 /0.53:
Số người sai (0 điểm): 17/52 (32.692%)
Số người đúng hoàn toàn (5.3 điểm): 35/52 (67.308%)

2. Q. 5 /0.53:
Số người sai (0 điểm): 15/52 (28.846%)
Số người đúng hoàn toàn (5.3 điểm): 37/52 (71.154%)

3. Q. 18 /0.53:
Số người sai (0 điểm): 14/52 (26.923%)
Số người đúng hoàn toàn (5.3 điểm): 38/52 (73.077%)

4. Q. 4 /0.53:
Số người sai (0 điểm): 13/52 (25.000%)
Số người đúng hoàn toàn (5.3 điểm): 39/52 (75.000%)

5. Q. 13 /0.53:
Số người sai (0 điểm): 11/52 (21.154%)
Số người đúng hoàn toàn (5.3 điểm): 41/52 (78.846%)

6. Q. 14 /0.53:
Số người sai (0 điểm): 11/52 (21.154%)
Số người đúng hoàn toàn (5.3 điểm): 41/52 (78.846%)

7. Q. 3 /0.53:
Số người sai (0 điểm): 10/52 (19.231%)
Số người đúng hoàn toàn (5.3 điểm): 42/52 (80.769%)

8. Q. 17 /0.53:
Số người sai (0 điểm): 9/52 (17.308%)
Số người đúng hoàn toàn (5.3 điểm): 43/52 (82.692%)

9. Q. 19 /0.53:
Số người sai (0 điểm): 8/52 (15.385%)
Số người

## Kết luận 2:

**Câu khó nhất: Q. 6 /0.53**  
- 17 sinh viên làm sai hoàn toàn - 0 điểm (32.692%)

---

**Câu dễ nhất: Q. 9 /0.53**  
- Không có sinh viên nào làm sai hoàn toàn
- 48 sinh viên làm đúng hoàn toàn - 5.3 điểm (92.308%)

---

**Câu mà có nhiều sinh viên kiếm được điểm tối đa nhiều nhất: Q. 16 /0.53**  
- Có 1 sinh viên làm sai hoàn toàn
- Còn lại thì tất cả 51 sinh viên làm đúng hoàn toàn - 5.3 điểm (98.077%)

## Câu hỏi 3
**Có mối tương quan giữa thời gian làm bài và điểm số không?**\
**Mục đích**: Tìm hiểu xem sinh viên làm bài nhanh hay chậm có ảnh hưởng đến điểm số không.

### Đơn giản
#### Bước 1: Tính thống kê cơ bản
- Tính thời gian trung bình của tất cả sinh viên
- Tính thời gian trung vị (median) - đây là mốc chia sinh viên thành 2 nhóm

#### Bước 2: Chia nhóm so sánh
- **Nhóm làm nhanh:** Sinh viên làm bài dưới thời gian trung vị
- **Nhóm làm chậm:** Sinh viên làm bài trên hoặc bằng thời gian trung vị
- Tính điểm trung bình của mỗi nhóm

#### Bước 3: Phân tích kết quả

**Có 3 tình huống có thể xảy ra:**

- **Nhóm làm nhanh có điểm cao hơn:** Có thể sinh viên giỏi nắm kiến thức vững nên làm nhanh và chính xác
- **Nhóm làm chậm có điểm cao hơn:** Có thể sinh viên cẩn thận, suy nghĩ kỹ sẽ đạt kết quả tốt hơn
- **Hai nhóm điểm tương đương:** Không có mối liên hệ rõ ràng giữa thời gian và điểm số

***Sinh viên hoàn thành tiếp trong các cell kế như bài tập về nhà***

### Bước 1: Tính thống kê cơ bản

In [ ]:
time_taken = cols['Time taken']
n = len(time_taken)
avg_time = sum(time_taken) / n

sorted_time = sorted(time_taken)
if n % 2 == 0:
    median_time = (sorted_time[n//2 - 1] + sorted_time[n//2]) / 2
else:
    median_time = sorted_time[n//2]

print(f"Thời gian trung bình: {avg_time:.3f} phút")
print(f"Thời gian trung vị: {median_time:.3f} phút")


Thời gian trung bình: 114.570 phút
Thời gian trung vị: 21.167 phút


### Bước 2: Chia nhóm so sánh

In [ ]:
fast_group = []
slow_group = []

for i in range(len(time_taken)):
    if time_taken[i] < median_time:
        fast_group.append(grades[i])
    else:
        slow_group.append(grades[i])

fast_avg = sum(fast_group) / len(fast_group)
slow_avg = sum(slow_group) / len(slow_group)

### Bước 3: Phân tích kết quả

In [18]:
print(f"\nNhóm làm nhanh (< {median_time:.3f} phút):")
print(f"- Số sinh viên: {len(fast_group)}")
print(f"- Điểm trung bình: {fast_avg:.3f}")

print(f"\nNhóm làm chậm (>= {median_time:.3f} phút):")
print(f"- Số sinh viên: {len(slow_group)}")
print(f"- Điểm trung bình: {slow_avg:.3f}")


Nhóm làm nhanh (< 21.167 phút):
- Số sinh viên: 26
- Điểm trung bình: 7.957

Nhóm làm chậm (>= 21.167 phút):
- Số sinh viên: 26
- Điểm trung bình: 9.008


## Kết luận:

- **Nhóm làm nhanh (< 21.167 phút):**
    - Số sinh viên: 26
    - Điểm trung bình: 7.957

- **Nhóm làm chậm (>= 21.167 phút):**
    - Số sinh viên: 26
    - Điểm trung bình: 9.008

---

**Nhóm làm bài nhanh có điểm trung bình hơn hơn 1.051 điểm so với nhóm làm bài chậm**  
$\to$ **Kết luận:** 
- Thời gian làm bài và điểm số có mối tương quan thuận  
- Sinh viên làm chậm (cẩn thận hơn) có xu hướng đạt điểm cao hơn.

### Nâng cao (không cần thư viện)
- Chia thời gian thành nhiều khoảng (ví dụ: 0-10 phút, 10-20 phút, 20-30 phút...)
- Tính điểm trung bình cho mỗi khoảng thời gian
- Quan sát xu hướng: điểm có tăng hay giảm theo thời gian không?

***Sinh viên hoàn thành tiếp trong các cell kế như bài tập về nhà***

### Bước 1: Chia thời gian thành nhiều khoảng

In [19]:
min_time = min(time_taken)
max_time = max(time_taken)

interval_size = 10
intervals = {}
current = 0
while current < max_time + interval_size:
    intervals[(current, current + interval_size)] = []
    current += interval_size

for i in range(len(time_taken)):
    time = time_taken[i]
    grade = grades[i]
    
    for (start, end) in intervals.keys():
        if start <= time < end:
            intervals[(start, end)].append(grade)
            break

### Bước 2: Tính điểm trung bình cho mỗi khoảng

In [20]:
# Tính điểm trung bình cho mỗi khoảng
print("Phân tích theo khoảng thời gian:")
for (start, end), grade_list in sorted(intervals.items()):
    if len(grade_list) > 0:
        avg_grade = sum(grade_list) / len(grade_list)
        print(f"[{start}-{end}) phút: Có {len(grade_list)} sinh viên, điểm trung bình = {avg_grade:.3f}")
    else:
        print(f"[{start}-{end}) phút: Không có sinh viên nào")

Phân tích theo khoảng thời gian:
[0-10) phút: Có 1 sinh viên, điểm trung bình = 9.740
[10-20) phút: Có 23 sinh viên, điểm trung bình = 7.896
[20-30) phút: Có 20 sinh viên, điểm trung bình = 8.981
[30-40) phút: Có 1 sinh viên, điểm trung bình = 10.000
[40-50) phút: Không có sinh viên nào
[50-60) phút: Có 2 sinh viên, điểm trung bình = 7.500
[60-70) phút: Có 1 sinh viên, điểm trung bình = 9.470
[70-80) phút: Có 1 sinh viên, điểm trung bình = 9.210
[80-90) phút: Không có sinh viên nào
[90-100) phút: Không có sinh viên nào
[100-110) phút: Không có sinh viên nào
[110-120) phút: Không có sinh viên nào
[120-130) phút: Có 1 sinh viên, điểm trung bình = 8.030
[130-140) phút: Có 1 sinh viên, điểm trung bình = 9.470
[140-150) phút: Không có sinh viên nào
[150-160) phút: Không có sinh viên nào
[160-170) phút: Không có sinh viên nào
[170-180) phút: Không có sinh viên nào
[180-190) phút: Không có sinh viên nào
[190-200) phút: Không có sinh viên nào
[200-210) phút: Không có sinh viên nào
[210-220) ph

## Phân tích xu hướng


### Nhận xét chi tiết:
- Giai đoạn **10–20 phút** có **nhiều sinh viên nhất (23 sinh viên)**, nhưng **điểm trung bình thấp hơn hẳn (7.90)**.  
- Các khoảng **thời gian dài hơn (20–40 phút, 60–80 phút, 130–140 phút)** thường có **điểm trung bình cao hơn mức điểm trung vị - 8.95**.  
- Một số khoảng rất ít sinh viên (chỉ 1–2 người), nên dữ liệu có thể **chưa đủ đại diện**, nhưng nhìn chung **điểm có xu hướng tăng khi thời gian làm bài dài hơn**.  
- Tuy nhiên, từ **50–60 phút** và **120–130 phút**, điểm trunh bình lại giảm (7.50 và 8.03), cho thấy **quá lâu có thể làm giảm hiệu suất** nếu không quản lý thời gian tốt.  

### Kết luận xu hướng:
- **Khi thời gian làm bài tăng từ 10 → 30 phút, điểm trung bình tăng rõ rệt** → *làm cẩn thận có lợi*.  
- **Khi vượt quá 60 phút**, điểm dao động, **không còn xu hướng tăng rõ ràng** → *ảnh hưởng của thời gian không ổn định*.  
- **Tổng thể**, xu hướng chung là:  
$\to$ *Làm bài quá nhanh dễ mất điểm, làm vừa phải (20–40 phút) đạt hiệu quả tốt nhất.*

## Câu hỏi 4
**Có sinh viên nào là điểm ngoại lệ (outliers) không?**\
**Mục đích:** Tìm những sinh viên có điểm số bất thường (quá cao hoặc quá thấp so với mức trung bình).

### Phương pháp 1: Sử dụng khoảng tứ phân vị (IQR - Interquartile Range)

#### Bước 1: Tính các giá trị thống kê
- Sắp xếp tất cả điểm số từ thấp đến cao
- Tính **Q1** (tứ phân vị thứ nhất): Điểm ở vị trí 25% (1/4 đầu dãy số)
- Tính **Q3** (tứ phân vị thứ ba): Điểm ở vị trí 75% (3/4 dãy số)
- Tính **IQR** = Q3 - Q1 (khoảng giữa)

**Ví dụ:** Có 100 sinh viên sau khi sắp xếp
- Q1 = điểm của sinh viên thứ 25 (ví dụ: 4.5)
- Q3 = điểm của sinh viên thứ 75 (ví dụ: 7.5)
- IQR = 7.5 - 4.5 = 3.0

#### Bước 2: Xác định ngưỡng outlier
- **Ngưỡng dưới** = Q1 - 1.5 × IQR
- **Ngưỡng trên** = Q3 + 1.5 × IQR

Tiếp ví dụ trên:
- Ngưỡng dưới = 4.5 - 1.5 × 3.0 = 0
- Ngưỡng trên = 7.5 + 1.5 × 3.0 = 12 (nhưng tối đa là 10)

#### Bước 3: Tìm outliers
- Sinh viên có điểm < ngưỡng dưới → **Outlier thấp** (điểm kém bất thường)
- Sinh viên có điểm > ngưỡng trên → **Outlier cao** (điểm xuất sắc bất thường)

***Sinh viên hoàn thành tiếp trong các cell kế như bài tập về nhà***

### Bước 1: Tính các giá trị thống kê

In [ ]:
sorted_grades = sorted(grades)
n = len(sorted_grades)

q1_index = int(0.25 * (n - 1))
Q1 = sorted_grades[q1_index]

q3_index = int(0.75 * (n - 1) + 1)
Q3 = sorted_grades[q3_index]

IQR = Q3 - Q1

print(f"Q1 (tứ phân vị thứ nhất): {Q1}")
print(f"Q3 (tứ phân vị thứ ba): {Q3}")
print(f"IQR (khoảng tứ phân vị): {IQR}")

Q1 (tứ phân vị thứ nhất): 7.11
Q3 (tứ phân vị thứ ba): 10.0
IQR (khoảng tứ phân vị): 2.8899999999999997


### Bước 2: Xác định ngưỡng outlier

In [22]:
lower_bound = max(0, Q1 - 1.5 * IQR)
upper_bound = min(10, Q3 + 1.5 * IQR)

print(f"\nNgưỡng dưới: {lower_bound:.3f}")
print(f"Ngưỡng trên: {upper_bound:.3f}")


Ngưỡng dưới: 2.775
Ngưỡng trên: 10.000


### Bước 2: Tìm outliers

In [23]:
low_outliers = []
high_outliers = []

for i, grade in enumerate(grades):
    if grade < lower_bound:
        low_outliers.append((i, grade))
    elif grade > upper_bound:
        high_outliers.append((i, grade))

if not low_outliers and not high_outliers:
    print("\nKhông có outlier theo phương pháp IQR")
else:
    print("\nCó outlier theo phương pháp IQR")
    
    print(f"\nSố sinh viên là outlier thấp: {len(low_outliers)}")
    if low_outliers:
        print("Danh sách (index, điểm):")
        for idx, grade in low_outliers:
            print(f"  Sinh viên thứ {idx+1}: {grade} điểm")

    print(f"\nSố sinh viên là outlier cao: {len(high_outliers)}")
    if high_outliers:
        print("Danh sách (index, điểm):")
        for idx, grade in high_outliers:
            print(f"  Sinh viên thứ {idx+1}: {grade} điểm")


Không có outlier theo phương pháp IQR


### Phân tích Outlier theo phương pháp IQR

- **Q1 (Tứ phân vị thứ nhất):** 7.11  
- **Q3 (Tứ phân vị thứ ba):** 10.00 (vì điểm số thang 10 nên max chỉ được 10)  
- **IQR (Khoảng tứ phân vị):** 2.89  

- **Ngưỡng dưới:** 2.7775  
- **Ngưỡng trên:** 10.00  

- **Số sinh viên là outlier thấp:** 0  
- **Số sinh viên là outlier cao:** 0  

$\to$ **Kết luận:** Không có outlier theo phương pháp IQR.


### Phương pháp 2: Sử dụng độ lệch chuẩn (Standard Deviation)

#### Bước 1: Tính giá trị trung bình và độ lệch chuẩn
- Tính điểm trung bình (mean)
- Tính độ lệch chuẩn (standard deviation) - đo mức độ phân tán của điểm số

#### Bước 2: Xác định ngưỡng
- **Ngưỡng dưới** = Mean - 2 × Std Dev
- **Ngưỡng trên** = Mean + 2 × Std Dev

**Ví dụ:**
- Điểm trung bình = 6.0
- Độ lệch chuẩn = 1.5
- Ngưỡng dưới = 6.0 - 2 × 1.5 = 3.0
- Ngưỡng trên = 6.0 + 2 × 1.5 = 9.0

#### Bước 3: Tìm outliers
- Điểm < 3.0 hoặc > 9.0 là outliers

***Sinh viên hoàn thành tiếp trong các cell kế như bài tập về nhà***

### Bước 1: Tính giá trị trung bình và độ lệch chuẩn

In [ ]:
mean = sum(grades) / len(grades)

variance_sum = 0
for grade in grades:
    variance_sum += (grade - mean) ** 2
variance = variance_sum / (len(grades) - 1)
std_dev = math.sqrt(variance)

print(f"Điểm trung bình (Mean): {mean:.3f}")
print(f"Độ lệch chuẩn (Std Dev): {std_dev:.3f}")

Điểm trung bình (Mean): 8.482
Độ lệch chuẩn (Std Dev): 1.699


### Bước 2: Xác định ngưỡng

In [25]:
lower_bound = mean - 2 * std_dev
upper_bound = mean + 2 * std_dev

print(f"\nNgưỡng dưới: {lower_bound:.3f}")
print(f"Ngưỡng trên: {upper_bound:.3f}")


Ngưỡng dưới: 5.084
Ngưỡng trên: 11.881


### Bước 3: Tìm outliers

In [26]:
low_outliers = []
high_outliers = []

for i, grade in enumerate(grades):
    if grade < lower_bound:
        low_outliers.append((i, grade))
    elif grade > upper_bound:
        high_outliers.append((i, grade))

if not low_outliers and not high_outliers:
    print("\nKhông có outlier theo phương pháp Std Dev")
else:
    print("\nCó outlier theo phương pháp Std Dev")
    
    print(f"\nSố sinh viên là outlier thấp: {len(low_outliers)}")
    if low_outliers:
        print("Danh sách (index, điểm):")
        for idx, grade in low_outliers:
            print(f"  Sinh viên thứ {idx+1}: {grade} điểm")

    print(f"\nSố sinh viên là outlier cao: {len(high_outliers)}")
    if high_outliers:
        print("Danh sách (index, điểm):")
        for idx, grade in high_outliers:
            print(f"  Sinh viên thứ {idx+1}: {grade} điểm")


Có outlier theo phương pháp Std Dev

Số sinh viên là outlier thấp: 2
Danh sách (index, điểm):
  Sinh viên thứ 5: 3.03 điểm
  Sinh viên thứ 41: 4.08 điểm

Số sinh viên là outlier cao: 0


### Phân tích Outlier theo phương pháp Độ lệch chuẩn (Standard Deviation)

- **Điểm trung bình (Mean):** 8.482  
- **Độ lệch chuẩn (Std Dev):** 1.699  

- **Ngưỡng dưới:** 5.084  
- **Ngưỡng trên:** 11.881  

$\to$ **Có outlier theo phương pháp Std Dev**

- **Số sinh viên là outlier thấp:** 2  
  - Sinh viên thứ **5**: 3.03 điểm  
  - Sinh viên thứ **41**: 4.08 điểm  

- **Số sinh viên là outlier cao:** 0